File: api/webhook.js
// webhook.js - Обработчик Telegram webhook для Vercel API routes
const https = require('https');

// Получаем токен бота из переменных окружения
const TELEGRAM_BOT_TOKEN = process.env.TEST_TELEGRAM_BOT_TOKEN;
const WEBHOOK_SECRET = process.env.TEST_WEBHOOK_SECRET;

// Базовый URL для Telegram API
const TELEGRAM_API_URL = `https://api.telegram.org/bot${TELEGRAM_BOT_TOKEN}`;

// Логирование с временной меткой
function log(level, message) {
  const timestamp = new Date().toISOString();
  console.log(`[${timestamp}] [${level}] ${message}`);
}

// Отправляет сообщение пользователю через Telegram API
async function sendTelegramMessage(chat_id, text, reply_markup = null) {
  return new Promise((resolve, reject) => {
    const payload = {
      chat_id: chat_id,
      text: text,
      parse_mode: "Markdown"
    };
    
    if (reply_markup) {
      payload.reply_markup = reply_markup;
    }
    
    const data = JSON.stringify(payload);
    
    log('INFO', `Отправка сообщения на ${chat_id}: ${text.substring(0, 50)}...`);
    
    const options = {
      hostname: 'api.telegram.org',
      port: 443,
      path: `/bot${TELEGRAM_BOT_TOKEN}/sendMessage`,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': data.length
      }
    };
    
    const req = https.request(options, (res) => {
      let responseData = '';
      
      res.on('data', (chunk) => {
        responseData += chunk;
      });
      
      res.on('end', () => {
        if (res.statusCode === 200) {
          log('INFO', `Сообщение успешно отправлено, chat_id: ${chat_id}`);
          resolve({ ok: true, response: JSON.parse(responseData) });
        } else {
          log('ERROR', `Ошибка API при отправке: ${res.statusCode}, ${responseData}`);
          resolve({ ok: false, error: `API error: ${res.statusCode}` });
        }
      });
    });
    
    req.on('error', (error) => {
      log('ERROR', `Исключение при отправке сообщения: ${error.message}`);
      reject({ ok: false, error: error.message });
    });
    
    req.write(data);
    req.end();
  });
}

// Обработчик запросов в формате Vercel API routes
export default async function handler(req, res) {
  try {
    // Логируем начало обработки запроса
    log('INFO', `Получен ${req.method} запрос на webhook`);
    
    // Обработка GET-запроса для проверки работоспособности
    if (req.method === 'GET') {
      return res.status(200).json({
        status: 'ok',
        message: 'Webhook активен',
        timestamp: new Date().toISOString(),
        version: '1.3.0'
      });
    }
    
    // Обработка POST-запроса от Telegram
    if (req.method === 'POST') {
      // Верификация запроса от Telegram
      const secretHeader = req.headers['x-telegram-bot-api-secret-token'];
      
      if (WEBHOOK_SECRET && secretHeader !== WEBHOOK_SECRET) {
        log('WARNING', "Получен неверифицированный запрос");
        return res.status(403).json({
          ok: false,
          error: 'Forbidden'
        });
      }
      
      // Получаем и парсим данные запроса
      const update = req.body;
      log('INFO', `Получено обновление ID: ${update.update_id}`);
      
      // Обработка текстовых сообщений
      if (update.message && update.message.text) {
        const chat_id = update.message.chat.id;
        const text = update.message.text;
        const user = update.message.from.first_name || 'пользователь';
        
        log('INFO', `Сообщение от ${user} (${chat_id}): ${text}`);
        
        if (text === '/start') {
          // Создаем клавиатуру с кнопками
          const keyboard = {
            "inline_keyboard": [
              [{"text": "О боте", "callback_data": "info"}],
              [{"text": "Тест", "callback_data": "test"}]
            ]
          };
          
          // Отправляем приветствие
          await sendTelegramMessage(
            chat_id, 
            `👋 Привет, ${user}!\n\nЯ тестовый бот для проекта DualAI на Vercel.\nЯ использую webhook-подход.`,
            keyboard
          );
        } else {
          // Отправляем эхо
          await sendTelegramMessage(chat_id, `Вы сказали: ${text}`);
        }
      }
      
      // Обработка нажатий на кнопки
      else if (update.callback_query) {
        const callback = update.callback_query;
        const chat_id = callback.message.chat.id;
        const data = callback.data;
        
        log('INFO', `Получен callback с данными: ${data}`);
        
        if (data === 'info') {
          await sendTelegramMessage(chat_id, "Это тестовый бот для проекта DualAI на Vercel через webhook.");
        } else if (data === 'test') {
          await sendTelegramMessage(chat_id, "Тестовое сообщение. Webhook работает!");
        } else {
          await sendTelegramMessage(chat_id, `Получена команда: ${data}`);
        }
      }
      
      // Отвечаем успехом, даже если не обработали конкретный тип сообщения
      return res.status(200).json({ ok: true });
    }
    
    // Обработка других HTTP методов
    return res.status(405).json({
      ok: false,
      error: 'Method not allowed'
    });
    
  } catch (error) {
    log('ERROR', `Общая ошибка обработки webhook: ${error.message}`);
    console.error(error.stack);
    
    return res.status(500).json({
      ok: false,
      error: error.message
    });
  }
} 

File: api/webhook.py
from http.server import BaseHTTPRequestHandler
import json
import os
import requests
import traceback
from datetime import datetime

# Получаем токен бота из переменных окружения
TELEGRAM_BOT_TOKEN = os.environ.get('TEST_TELEGRAM_BOT_TOKEN')
if not TELEGRAM_BOT_TOKEN:
    TELEGRAM_BOT_TOKEN = "7513434644:AAECYxIDIkmZRjGgUDrP8ur2cZIni53Qy0E"  # Fallback токен

# Базовый URL для Telegram API
TELEGRAM_API_URL = f"https://api.telegram.org/bot{TELEGRAM_BOT_TOKEN}"

def send_message(chat_id, text, reply_markup=None):
    """Отправляет сообщение пользователю через Telegram API."""
    url = f"{TELEGRAM_API_URL}/sendMessage"
    payload = {
        "chat_id": chat_id,
        "text": text,
        "parse_mode": "Markdown"
    }
    
    if reply_markup:
        payload["reply_markup"] = json.dumps(reply_markup)
    
    try:
        print(f"Отправка сообщения на {chat_id}: {text}")
        response = requests.post(url, json=payload)
        print(f"Ответ API: {response.text}")
        return {"ok": True, "response": response.text}
    except Exception as e:
        print(f"Ошибка отправки сообщения: {e}")
        traceback.print_exc()
        return {"ok": False, "error": str(e)}

class handler(BaseHTTPRequestHandler):
    def do_GET(self):
        print(f"GET запрос на {self.path}")
        self.send_response(200)
        self.send_header('Content-type', 'application/json')
        self.end_headers()
        
        response_data = {
            'status': 'ok',
            'message': 'Webhook активен',
            'timestamp': str(datetime.now())
        }
        
        self.wfile.write(json.dumps(response_data).encode())
        return
    
    def do_POST(self):
        print(f"POST запрос на {self.path}")
        
        try:
            content_length = int(self.headers['Content-Length'])
            post_data = self.rfile.read(content_length)
            update = json.loads(post_data.decode('utf-8'))
            
            print(f"Получены данные: {json.dumps(update)}")
            
            # Готовим ответ на webhook
            self.send_response(200)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            
            response_data = {'status': 'ok'}
            self.wfile.write(json.dumps(response_data).encode())
            
            # После отправки ответа на webhook обрабатываем сообщение
            try:
                # Проверяем наличие сообщения с текстом
                if 'message' in update and 'text' in update['message']:
                    chat_id = update['message']['chat']['id']
                    message_text = update['message']['text']
                    user_name = update['message']['from'].get('first_name', 'пользователь')
                    
                    print(f"Сообщение от {user_name} ({chat_id}): {message_text}")
                    
                    # Обработка команды /start
                    if message_text == '/start':
                        reply_text = f"Привет, {user_name}! Я тестовый бот для проекта DualAI."
                        
                        # Создаем клавиатуру
                        keyboard = {
                            "inline_keyboard": [
                                [{"text": "Информация", "callback_data": "info"}],
                                [{"text": "Тест", "callback_data": "test"}]
                            ]
                        }
                        
                        # Отправляем сообщение с клавиатурой
                        send_result = send_message(chat_id, reply_text, keyboard)
                        print(f"Результат отправки: {json.dumps(send_result)}")
                    else:
                        # Эхо для других сообщений
                        reply_text = f"Вы сказали: {message_text}"
                        send_result = send_message(chat_id, reply_text)
                        print(f"Результат отправки: {json.dumps(send_result)}")
                
                # Обработка нажатий на кнопки
                elif 'callback_query' in update:
                    callback_query = update['callback_query']
                    chat_id = callback_query['message']['chat']['id']
                    data = callback_query['data']
                    
                    print(f"Callback query с данными: {data}")
                    
                    if data == 'info':
                        text = "Это тестовый бот для проекта DualAI на Vercel."
                    else:
                        text = "Тестовый ответ на callback."
                    
                    send_result = send_message(chat_id, text)
                    print(f"Результат отправки: {json.dumps(send_result)}")
            except Exception as e:
                print(f"Ошибка при обработке сообщения: {e}")
                traceback.print_exc()
            
            return
        except Exception as e:
            print(f"Общая ошибка обработки webhook: {e}")
            traceback.print_exc()
            
            self.send_response(500)
            self.send_header('Content-type', 'application/json')
            self.end_headers()
            
            response_data = {'status': 'error', 'error': str(e)}
            self.wfile.write(json.dumps(response_data).encode())
            return 